trace/sns/message.go

```diff
@@ -0,0 +1,182 @@
+package sns
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/service/sns"
+	"github.com/beatlabs/patron/errors"
+)
+
+type attributeDataType string
+
+const (
+	attributeDataTypeString      attributeDataType = "String"
+	attributeDataTypeStringArray attributeDataType = "StringArray"
+	attributeDataTypeNumber      attributeDataType = "Number"
+	attributeDataTypeBinary      attributeDataType = "Binary"
+
+	tracingTargetTopicArn    = "topic-arn"
+	tracingTargetTargetArn   = "target-arn"
+	tracingTargetPhoneNumber = "phone-number"
+	tracingTargetUnknown     = "unknown"
+)
+
+// MessageBuilder is a builder that helps building messages to be sent to SNS.
+type MessageBuilder struct {
+	err   error
+	input *sns.PublishInput
+}
+
+// NewMessageBuilder creates a new MessageBuilder that helps creating messages.
+func NewMessageBuilder() *MessageBuilder {
+	return &MessageBuilder{
+		input: &sns.PublishInput{
+			MessageAttributes: map[string]*sns.MessageAttributeValue{},
+		},
+	}
+}
+
+// Message is a struct embedding information about messages that will
+// be later published to SNS thanks to the SNS publisher.
+type Message struct {
+	input *sns.PublishInput
+}
+
+// tracingTarget returns a string used for tracing operations. As a message can only define one TopicArn,
+// TargetArn or PhoneNumber, we set a different tracing topic for each case.
+func (m Message) tracingTarget() string {
+	if m.input.TopicArn != nil {
+		return fmt.Sprintf("%s:%s", tracingTargetTopicArn, *m.input.TopicArn)
+	}
+
+	if m.input.TargetArn != nil {
+		return fmt.Sprintf("%s:%s", tracingTargetTargetArn, *m.input.TargetArn)
+	}
+
+	if m.input.PhoneNumber != nil {
+		// We don't append the phone number so that we don't end with 1 target per phone number.
+		return tracingTargetPhoneNumber
+	}
+
+	return tracingTargetUnknown
+}
+
+// WithMessage attaches a message to the message struct.
+func (b *MessageBuilder) WithMessage(msg string) *MessageBuilder {
```

Comment:
I am not sure about the builder here because there is not that much optional to configure here.
An SNS payload has to have a message, a topic arn etc.
We should have the required ones not as options with "With***".